addons/addon-internal-auth-dep-check/packages/internal-auth-dep-check/lib/steps/internal-auth-dep-check-service.js

```diff
@@ -0,0 +1,255 @@
+    }
+
+    // Verify all internal users (including root) are deactivated
+    if (!(await this.verifyNoActiveInternalUsers(listOfInternalStatuses))) {
+      throw this.boom.badRequest('Please deactive all Internal IdP users before upgrading');
+    }
+  }
+
+  async verifyInternalUserWorkspacesAreTerminated(listOfInternalUsers) {
+    const table = this.settings.get(settingKeys.environemntsTableName);
+
+    // Scan for any environments that are not in a failed or terminated state
+    const result = await this.dynamoDB
+      .scanner()
+      .table(table)
+      .names({ '#s': 'status' })
+      .values({ ':c': 'COMPLETED', ':p': 'PENDING', ':start': 'STARTING', ':stop': 'STOPPING', ':t': 'TERMINATING' })
+      .filter('#s = :c OR #s = :p OR #s = :start OR #s = :stop OR #s = :t')
+      .projection(['id', 'createdBy'])
+      .scan();
+
+    const listOfNonTerminatedWorkspaces = _.uniq(
+      result.filter(item => listOfInternalUsers.includes(item.createdBy)).map(item => item.id),
+    );
+
+    if (listOfNonTerminatedWorkspaces.length > 0) {
+      this.log.error(`Found at least one workspace that is not terminated: ${listOfNonTerminatedWorkspaces}`);
+      return false;
+    }
+
+    return true;
+  }
+
+  async verifyNoInternalUserSSHKey(listOfInternalUsers) {
+    const table = this.settings.get(settingKeys.keyPairsTableName);
+
+    // Scan for active ssh keys
+    const result = await this.dynamoDB
+      .scanner()
+      .table(table)
+      .names({ '#s': 'status' })
+      .values({ ':a': 'active' })
+      .filter('#s = :a')
+      .projection('uid')
+      .scan();
+
+    const listOfUsers = _.uniq(result.filter(item => listOfInternalUsers.includes(item.uid)).map(item => item.uid));
+
+    if (listOfUsers.length > 0) {
+      this.log.error(`Found at least one user with active SSH key: ${listOfUsers}`);
+      return false;
+    }
+    return true;
+  }
+
+  async _listInternalUsers() {
+    const table = this.settings.get(settingKeys.usersTableName);
+
+    // Scan for internal authentication users
+    const result = await this.dynamoDB
+      .scanner()
+      .table(table)
+      .names({ '#a': 'authenticationProviderId' })
+      .values({ ':i': 'internal' })
+      .filter('#a = :i')
+      .projection(['uid', 'status'])
+      .scan();
+
+    const listOfInternalUsers = _.uniq(result.map(item => item.uid));
+    const listOfInternalStatuses = _.uniq(result.map(item => item.status));
+
+    return { listOfInternalUsers, listOfInternalStatuses };
+  }
+
+  async verifyNoActiveInternalUsers(listOfInternalStatuses) {
+    if (listOfInternalStatuses.includes('active')) {
+      this.log.error(`Found an least one active internal user`);
+      return false;
+    }
+    return true;
+  }
+
+  async verifyNoInternalUserProjects(listOfInternalUsers) {
+    const table = this.settings.get(settingKeys.projectsTableName);
+
+    // Scan for project admins
```

Comment:
The project admins feature is not implemented yet.

This method should check that internal users are not assigned a project (instead of project assigned to a user). 

The reason we want to do this is that if in the future, IdP users accidentally activate an internal user, then these deprecated users will show up in user dropdowns throughout the application just because they are still linked to the respective projectId. This might cause confusion for the admin assigning the IdP user vs the deprecated internal user (since their names will be the same).

eg. Admin is trying to assign user John Doe to resourceABC in TestProj, but there are two John Doe's in the dropdown (one internal user, and one IdP), both assigned to TestProj.

Reply:
Got it! Thanks for the clarification!